### PR TITLE
Refactor `RustTarget`

### PR DESCRIPTION
**Pull request by pvdrz**
This PR refactors the `RustTarget` type to be more flexible.

The motivation behind this change comes from the [current definition](https://docs.rs/bindgen/0.70.1/bindgen/enum.RustTarget.html) of `RustTarget` as an enum. Which forces users to pick a specific version of Rust that might not match the msrv of their projects exactly.

As an example, let's say that a project using bindgen has an msrv of `1.41.1` and they decide to  use the `--rust-target` flag to guarantee that bindgen won't produce code that breaks their msrv. 

Given that bindgen doesn't have a `RustTarget` variant that specifically matches `1.41.1`, they have to use the closest `RustTarget` that is still compatible. In this case that would be `1.40`. If they decide to change their msrv in the future, they have to repeat this search process with their new msrv.

With the changes done in this PR, it would be possible to pass any* Rust version to `--rust-target` and bindgen will automatically enable the Rust features that are compatible with that version. In our example, that means that it would be possible to use `--rust-version=1.41.1` explicitly, or even better, read the value from `rust-version` so they don't have to manually change it in the future.

The specific changes done in this PR are:
- Replace the `RustTarget` enum for an opaque struct that can represent any* Rust version.
- Update the `FromStr` implementation for `RustTarget` so it can parse any* Rust version.
- Introduce several deprecated constants under `RustTarget` so people can still use those as if they were still using the variants of the removed enum type.
- Introduce the builders `RustTarget::nightly` and `RustTarget::stable`.
- Remove all the deprecated `RustTarget`s.

- Update the expectation tests to accommodate the removal of the deprecated targets.

(*): Unless it's not supported by bindgen anymore. To the date, anything earlier than 1.33 (inclusive)

cc @ojeda

### CODE REVIEW

**Comment by ojeda** [comment]
Thanks for this, Christian!

Just to confirm: for projects that need to support `bindgen` releases with and without this PR, it is still possible to pass their current (old) `--rust-target`, i.e. all previous valid inputs will still be valid ones with the PR and do the same thing, right? (I see the parsing allows minor only, i.e. `1.x`, so I think that is the intention, which is great.)

**Comment by pvdrz** [comment]
> Thanks for this, Christian!
> 
> Just to confirm: for projects that need to support `bindgen` releases with and without this PR, it is still possible to pass their current (old) `--rust-target`, i.e. all previous valid inputs will still be valid ones with the PR and do the same thing, right? (I see the parsing allows minor only, i.e. `1.x`, so I think that is the intention, which is great.)

Yes all the previous valid inputs are still valid, e.g. `1.33` gets parsed as `1.33.0` and so on. The only breaking change I expect to see here (other than the removal of deprecated targets) is if anyone was trying to `match` the `RustTarget` enum as their matches would no longer be exhaustive

**Comment by ojeda** [comment]
Sounds great, thanks for confirming!

The `RustTarget` part wouldn't affect the kernel since we use the CLI so far. Regarding the `match`, it is true that they wouldn't be able to simply match all without a wildcard; however, in that case, they needed to add new variants anyway when new releases added a new target version, no?

File: bindgen/features.rs [hunk 1]
```diff
@@ -170,6 +170,42 @@ pub const LATEST_STABLE_RUST: RustTarget = {
...
+    //     .min_by_key(|(_, m)| m)
+    //     .map(|(t, _)| t)
+    //     .unwrap_or(LATEST_STABLE_RUST)
+    // ```
+    // once those operations can be used in constants.
+    let targets = RustTarget::stable_releases();
+
+    let mut i = 0;
+    let mut earliest_target = None;
+    let Some(mut earliest_minor) = LATEST_STABLE_RUST.minor() else {
```
Comment (emilio) [inline bindgen/features.rs:225]: Maybe `LATEST_STABLE_RUST.minor().expect("No minor version?")`? I guess it has to be this way for `const`-ness?

File: bindgen-tests/tests/headers/layout.h [hunk 2]
```diff
@@ -1,10 +0,0 @@
-// bindgen-flags: \-\-rust-target=1.33
-//
-// FIXME: https://github.com/rust-lang/rust-bindgen/issues/1498
```
Comment (emilio) [inline bindgen-tests/tests/headers/layout.h:3]: Rather than removing this one, let's keep it with `#if 0` or so?

**Review by emilio — APPROVED** [review]
Would've been nice to send the removal of deprecated targets separately, but looks good! Some minor bits only.

File: bindgen/features.rs (hunk 1 above)
Comment (pvdrz) [reply bindgen/features.rs:225]: yeah it is for const-ness

**Pull request merged**